### PR TITLE
feat: add 4 dimensional vector object

### PR DESCRIPTION
**Pull request by osbornjd**
BEGINRELEASENOTES
- Added a Vector4f object for use as a general 4vector. Addresses https://github.com/key4hep/EDM4hep/issues/245

ENDRELEASENOTES

### CODE REVIEW

File: edm4hep.yaml [hunk 1]
```diff
@@ -6,6 +6,23 @@ options:
   includeSubfolder: True
 
 components:
+    
+    #Vector4f with floats
+    edm4hep::Vector4f:
+      Description: "Generic 4 vector for storing classical 4D coordinates in memory. 4 vector helper functions are in edm4hep::utils"
```
Comment (tmadlener) [inline edm4hep.yaml:12]:
```suggestion
      Description: "Generic vector for storing classical 4D coordinates in memory. Four momentum helper functions are in edm4hep::utils"
```
Maybe we can avoid some of the confusion by calling this a 4D vector and refer to the Lorentz vector as four momentum (and then just try to do that consistently)

File: edm4hep.yaml [hunk 2]
```diff
@@ -6,6 +6,23 @@ options:
   includeSubfolder: True
 
 components:
+    
+    #Vector4f with floats
```
Comment (tmadlener) [inline edm4hep.yaml:10]:
```suggestion
```
I think that is quite obvious from the lines below.

**Review by tmadlener — COMMENTED** [review]
Thanks. I think this looks good with a few minor corrections. 

There are some vector utilities that have been added by EIC here: [`utils/include/edm4hep/utils/vector_utils.h`](https://github.com/key4hep/EDM4hep/blob/main/utils/include/edm4hep/utils/vector_utils.h). I am not sure if you also want to add this one there, or if you want to wait with that until there is a need to do it.

Usually `pre-commit` will fail because the `README.md` has links to the different types in the YAML file and it checks whether they point to the right place. You can easily fix this via the `scripts/updateReadmeLinks.py` script.

**Comment by osbornjd** [comment]
> There are some vector utilities that have been added by EIC here: [utils/include/edm4hep/utils/vector_utils.h](https://github.com/key4hep/EDM4hep/blob/main/utils/include/edm4hep/utils/vector_utils.h)

I'll work on implementing some of these, as they would probably be useful in the long run

File: edm4hep.yaml [hunk 3]
```diff
@@ -6,6 +6,22 @@ options:
   includeSubfolder: True
 
 components:
+
```
Comment (tmadlener) [inline edm4hep.yaml:9]:
```suggestion

```

**Comment by osbornjd** [comment]
I added some basic operator functionality for the 4 vectors. I think that should be sufficient for now and if we need more functionality, I can come back and add it in a future PR.

**Comment by osbornjd** [comment]
What is the best way to test the build failure? This builds for me locally in my environment and all tests pass, so I'm not sure how best to debug. I did re-test with the `precommit run --all-files` and these seem to pass now

**Comment by tmadlener** [comment]
I think the build failures are in the c+17 builds. There is a [`vector_utils_legacy.h`](https://github.com/key4hep/EDM4hep/blob/main/utils/include/edm4hep/utils/vector_utils_legacy.h) that get's dropped in in case there are no concepts available yet:

https://github.com/key4hep/EDM4hep/blob/e48126265a1176ab02dfae472633e94869dfb4e5/utils/include/edm4hep/utils/vector_utils.h#L6-L8

So you would have to implement at least the `vector_XYZ` specializations for `Vector4f` there.

**Comment by osbornjd** [comment]
Looks like everything passes except that a test in an fcc build seg faults. What is the best way for me to test this locally as I don't see these tests in the repo?

**Comment by tmadlener** [comment]
That seems to be unrelated to this change and I am not entirely sure whether the `downstream` workflow is already supposed to be working. It has been a bit finicky in the past.

**Comment by osbornjd** [comment]
Okay sounds good, thanks for letting me know. In that case, if it is not too much of an issue would you mind creating a release after merging this PR? It is not urgent with the holidays coming up, so there is absolutely no rush on this

**Review by tmadlener — APPROVED** [review]

**Comment by tmadlener** [comment]
I will wait with merging until after the break even though this should not break anything. If nothing has happened after a few days in January, can you ping me again? A new tag is no problem.

**Comment by osbornjd** [comment]
Sounds great, will do. Thanks for your help!

**Comment by osbornjd** [comment]
Unless there are any further changes requested (and now since it is after the holiday break), are there any objections to merge this and create a new release?

**Review by tmadlener — APPROVED** [review]

**Comment by tmadlener** [comment]
There is a new tag and there is also a PR in spack to pick it up: https://github.com/spack/spack/pull/41987

**Comment by osbornjd** [comment]
Thanks very much for the new tag!

**Comment by wdconinc** [comment]
Merged in spack.

**Pull request merged**